### PR TITLE
Add user cmake support

### PR DESCRIPTION
**Pull request by xantares**
Addresses #80: allow USERS to build fftw (not the developer side).

Supports static/shared, openmp, threads, various precision options, sse/avx.
Does not support fortran headers, mpi, other processor arches. Maybe that can added in the future. I think it's reasonably complete for now.

Tested here on linux, osx, msvc: https://github.com/xantares/fftw-cmake

cc @EikeVerdenhalven @junghans

### CODE REVIEW

File: cmake.config.h.in [hunk 1]
```diff
@@ -0,0 +1,429 @@
+/* config.h.  Generated from config.h.in by configure.  */
+/* config.h.in.  Generated from configure.ac by autoheader.  */
```
Comment (stevengj) [inline cmake.config.h.in]: Can this file be automatically generated, e.g. by adding a few lines to `bootstrap.sh`?

Comment (stevengj) [inline cmake.config.h.in]: If it can't be automatically generated, then the header to the file should be changed.  And probably the `#defines` that are unrelated to supported cmake platforms should be deleted.

File: cmake.config.h.in [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+#define F77_FUNC_(name,NAME) name ## _
+
+/* Define if F77_FUNC and F77_FUNC_ are equivalent. */
+#define F77_FUNC_EQUIV 1
+
+/* Define if F77 and FC dummy `main' functions are identical. */
+/* #undef FC_DUMMY_MAIN_EQ_F77 */
+
+/* C compiler name and flags */
+#define FFTW_CC "gcc -std=gnu99 -O3 -fomit-frame-pointer -mtune=native -malign-double -fstrict-aliasing -fno-schedule-insns -ffast-math"
```
Comment (stevengj) [inline cmake.config.h.in]: It really seems wrong to hard-code this.

File: cmake.config.h.in [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+/* #undef HAVE_BSDGETTIMEOFDAY */
+
+/* Define to 1 if you have the `clock_gettime' function. */
+#cmakedefine HAVE_CLOCK_GETTIME
+
+/* Define to 1 if you have the `cosl' function. */
+#define HAVE_COSL 1
+
+/* Define to 1 if you have the <c_asm.h> header file. */
+#cmakedefine HAVE_C_ASM_H 1
```
Comment (stevengj) [inline cmake.config.h.in]: I thought that `cmake` can check for headers etc?

File: cmake.config.h.in [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+
+/* Version number of package */
+#define VERSION "@FFTW_VERSION@"
+
+/* Use common Windows Fortran mangling styles for the Fortran interfaces. */
+/* #undef WINDOWS_F77_MANGLING */
+
+/* Include g77-compatible wrappers in addition to any other Fortran wrappers.
+   */
+#define WITH_G77_WRAPPERS 1
```
Comment (stevengj) [inline cmake.config.h.in]: Not `#cmakedefine`?

File: cmake.config.h.in (hunk 3 above)
Comment (xantares) [inline cmake.config.h.in]: The CMakeLists checks for c_asm.h, then depending on the value of HAVE_C_ASM_H in the CmakeLists the config.h is configured to define or not HAVE_C_ASM_H.

Comment (stevengj) [inline cmake.config.h.in]: What about things like `HAVE_COSL`?

Comment (stevengj) [inline cmake.config.h.in]: Or is the cmake build only intended to target platforms that have `cosl` and other C99 things?

File: cmake.config.h.in (hunk 4 above)
Comment (xantares) [inline cmake.config.h.in]: The fortran part is not supported yet.

File: cmake.config.h.in (hunk 3 above)
Comment (xantares) [inline cmake.config.h.in]: I just added cosl/sinl detection.

File: cmake.config.h.in (hunk 4 above)
Comment (xantares) [inline cmake.config.h.in]: added

File: cmake.config.h.in (hunk 2 above)
Comment (xantares) [inline cmake.config.h.in]: ok, fixed

File: cmake.config.h.in (hunk 1 above)
Comment (xantares) [inline cmake.config.h.in]: I dont think it can be automatically generated easily

File: cmake.config.h.in [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+#define HAVE_VPRINTF 1
+
+/* Define to enable IBM VSX optimizations. */
+/* #undef HAVE_VSX */
+
+/* Define to 1 if you have the `_mm_free' function. */
+/* #undef HAVE__MM_FREE */
+
+/* Define to 1 if you have the `_mm_malloc' function. */
+/* #undef HAVE__MM_MALLOC */
```
Comment (stevengj) [inline cmake.config.h.in]: The `_mm_malloc` and `_mm_free` functions are useful when the code is compiled [with the Intel compiler](https://software.intel.com/en-us/articles/memory-management-for-optimal-performance-on-intel-xeon-phi-coprocessor-alignment-and), which presumably we want to support in cmake.

File: cmake.config.h.in [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+#define SIZEOF_FFTW_R2R_KIND 4
+
+/* The size of `float', as computed by sizeof. */
+#define SIZEOF_FLOAT 4
+
+/* The size of `int', as computed by sizeof. */
+#define SIZEOF_INT 4
+
+/* The size of `long', as computed by sizeof. */
+#define SIZEOF_LONG 8
```
Comment (stevengj) [inline cmake.config.h.in]: This is incorrect for Windows, for example.  Presumably these should be determined by [CheckTypeSize](https://cmake.org/cmake/help/v3.0/module/CheckTypeSize.html) in cmake.

File: cmake.config.h.in (hunk 5 above)
Comment (stevengj) [inline cmake.config.h.in]: Actually, it should be fine to omit these, since we [check explicitly for icc](https://github.com/FFTW/fftw3/blob/a71f3dd355f802dc362a52674a977ff81daadf9d/kernel/kalloc.c#L100-L104) even if `HAVE__MM_MALLOC` is not defined.

Comment (xantares) [inline cmake.config.h.in]: ok

File: cmake.config.h.in (hunk 6 above)
Comment (xantares) [inline cmake.config.h.in]: ok, done

**Comment by xantares** [comment]
what do you say guys ?

**Comment by matteo-frigo** [comment]
Alright, I merged your changes and made some fixes.

There is still a lot of cruft may not make any sense to keep around with cmake.  For example, c_asm.h was necessary I think on Digital Unix on alpha.  Our autoconf build scripts will of course support this configuration forever, but we should decide what are we targeting with cmake.  If we are targeting only windows or windows/linux/osx, a lot of junk can go away.

**Comment by matteo-frigo** [comment]
Is HAVE_DLFCN_H necessary?  We don't use it at all.

**Comment by matteo-frigo** [comment]
What's the best way for me to test that this works on Windows?  I never use Windows, so I need exact instructions of the form: "download compiler XXX from http://YYY, download cmake from ZZZ, run cmake in this way: QQQ"

**Comment by matteo-frigo** [comment]
@xantares One major flaw in this build system is that it is impossible to create a portable binary.  If you enable AVX, everything is compiled with -mavx, and the binary will crash when run on a processor without AVX.  The autoconf build system only uses -mavx in the */simd/avx subdirectories, and those are only called after a runtime check that AVX is supported by the CPU.

Is there any way to achieve the same effect in cmake?  This is pretty important: with the current zoo of simd extensions, runtime detection is necessary.

**Comment by junghans** [comment]
@xantares: Instead of `target_compile_options (${fftw3_lib} PRIVATE ${AVX_FLAG})`, use `set_source_files_properties(<file> COMPILE_FLAGS ${AVX_FLAG})` to make it on per-file basis.

**Comment by matteo-frigo** [comment]
@xantares @junghans I finally managed to use the cmakefile with VS2017, and it kind of works but not too well yet. 

First, the /ARCH autodetection is not right.  Apparently, VS2017 does not recognize /ARCH:SSE2 in x64 mode (since SSE2 is required in 64-bit mode), and so cmake disables sse2 code, which is wrong.  Moreover, in x86 mode, one cannot specify /ARCH:AVX2 and /ARCH:AVX, since AVX2 subsumes AVX, and so we get warnings.  Any suggestions?  I am inclined to remove autodetection completely and let users specify /ARCH by hand.

Second, performance seems to be substandard.  Admittedly I have only ran one test in a VM, but on the same hardware, gcc runs a 1k transform at 27.5 GFLOPS, and VS runs at ~20GFLOPS.  So the out-of-the-box experience does not seem right.  We'll need to worry about this once the /ARCH business is fixed.

**Comment by matteo-frigo** [comment]
@xantares @junghans Any suggestion on the /ARCH autodetection?  I would like to make a release at some point.
